### ISSUE TITLE
Same function, same arguments, different generated code

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#20913]
Consider the following small fixedsize matrix implementation together with a reduce function and two implementations of `sum`:

```jl
struct FixedMatrix{R,C,T,RC} <: AbstractMatrix{T}
    data::NTuple{RC, T}
end

Base.size{R,C,T,RC}(::FixedMatrix{R,C,T,RC}) = (R,C)
Base.length{R,C,T,RC}(::FixedMatrix{R,C,T,RC}) = RC
Base.IndexStyle(::Type{<: FixedMatrix}) = IndexLinear()
Base.getindex(fm::FixedMatrix, i::Int) = fm.data[i]

@inline function myreduce(op, v0, a::FixedMatrix)
    if length(a) == 0
        return v0
    else
        s = v0
        @inbounds @simd for j = 1:length(a)
            s = op(s, a[j])
        end
        return s
    end
end

sum2(a::FixedMatrix) = myreduce(+, zero(eltype(a)), a)
sum3(a::FixedMatrix) = myreduce(+, 0.0, a)
```

Now, `sum2` and `sum3` will (should?) clearly call the same method with the same arguments when the matrix is of type `Float64` (which I confirmed by putting a `@which` in there).

However:

```jl
julia> using BenchmarkTools

julia> m = FixedMatrix{8, 8, Float64, 64}((rand(64)...))

julia> @btime sum2($m)
  67.113 ns (0 allocations: 0 bytes)
36.441929014270464

julia> @btime sum3($m)
  8.379 ns (0 allocations: 0 bytes)
36.441929014270464
```

In fact, the generated code is vastly different (one uses SIMD, the other one does not).

```jl
julia> @code_llvm sum2(m)

julia> @code_llvm sum3(m)
```

How can so different code be generated when the exact same function is called with the same arguments?

**Comment by KristofferC** [comment]
Also not that the built in sum time is slower than the one I wrote up there:

```
julia> @btime sum($m)
  17.394 ns (0 allocations: 0 bytes)
36.441929014270464
```

**Comment by yuyichao** [comment]
What's `m`?

**Comment by KristofferC** [comment]
Sorry! `m = FixedMatrix{8, 8, Float64, 64}((rand(64)...))`

**Comment by fredrikekre** [comment]
```julia
julia> @btime sum2($m);
  9.603 ns (0 allocations: 0 bytes)

julia> @btime sum3($m);
  9.341 ns (0 allocations: 0 bytes)
```

**Comment by KristofferC** [comment]
What version? Is the generated code the same for you?

**Comment by KristofferC** [comment]
My master was 11 days old... maybe it has been fixed already. Will look.

**Comment by fredrikekre** [comment]
I got the same generated code but for a 5x5 case. See my edit in the comment above.

**Comment by KristofferC** [comment]
Yes for 5x5 I think everything gets unrolled.

**Comment by KristofferC** [comment]
Same result on 0.5.

**Comment by JeffBezanson** [comment]
The only difference in IR seems to be a constant 0.0 vs. `sitofp(0)`. Maybe the `sitofp` is enough to throw off the optimizer? I guess we should add some early constant folding to `sitofp`.

**Comment by Keno** [comment]
That would be either a serious bug in LLVM or a pass order problem. Either way we should find out.

**Comment by KristofferC** [comment]
Any ideas here?

**Comment by KristofferC** [comment]
Still happens

**Comment by KristofferC** [comment]
This seems to be fixed on 0.7 :tada:. 

```
julia> @btime sum2($m)
  7.670 ns (0 allocations: 0 bytes)
35.36299421828035

julia> @btime sum3($m)
  7.603 ns (0 allocations: 0 bytes)
35.36299421828035
```
